File: course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        List[] neighbors = new ArrayList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            neighbors[i] = new ArrayList<Integer>();
        }
        for (int[] pre : prerequisites) {
            int in = pre[0];
            int from = pre[1];
            indegree[in]++;
            neighbors[from].add(in);
        }
        List<Integer> res = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
            }
        }
        while (!queue.isEmpty()) {
            int course = queue.poll();
            res.add(course);
            List<Integer> neighbor = neighbors[course];
            for (int n : neighbor) {
                indegree[n]--;
                if (indegree[n] == 0) {
                    queue.offer(n);
                }
            }
        }
        if (res.size() != numCourses) {
            return new int[0];
        }
        int[] ans = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            ans[i] = res.get(i);
        }
        return ans;
    }
}

File: course-schedule-ii.py
import collections

class Solution:
    def findOrder(self, numCourses, prerequisites):
        res = []
        indegree = [0] * numCourses
        neighbors = [[] for _ in range(numCourses)]
        for dst, src in prerequisites:
            indegree[dst] += 1
            neighbors[src].append(dst)
        
        queue = collections.deque([i for i, n in enumerate(indegree)if n == 0])
        while queue:
            node = queue.popleft()
            res.append(node)
            for neighbor in neighbors[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)

        return res if len(res) == numCourses else []
